### PR TITLE
#40 Sender instance passed to the Factory

### PR DESCRIPTION
**Pull request by danbilauca**
I've chosen to implement a Sender Factory and send it along to Factory's constructor in order to have code consistency and have it processed/validated by Factory's logic.

### CODE REVIEW

File: includes/Factory.php [hunk 1]
```diff
@@ -19,28 +19,45 @@ class WPNotify_Factory {
...
+	 *
+	 * @param $args
+	 *
+	 * @return WPNotify_BaseNotification
+	 */
 	public function create( $args ) {
-		list( $message_args, $recipients_args ) = $this->validate( $args );
 
-		$message = $this->message_factory->create( $message_args );
+		[ $message_args, $recipients_args, $sender_args ] = $this->validate( $args );
```
Comment (raaaahman) [inline includes/Factory.php]: Short array syntax is not in the WordPress coding standards...

We should configure the PHP Code Sniffer to help us with these standards.

File: includes/senders/SenderFactory.php [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+interface WPNotify_SenderFactory {
+
+	/**
+	 * Create a new instance of notification sender
+	 *
+	 * @param string $name
+	 *
+	 * @return WPNotify_Sender
+	 */
+	public function create( string $name ): WPNotify_Sender;
```
Comment (raaaahman) [inline includes/senders/SenderFactory.php]: I'm not sure if we should use static type hinting or not.

What would be the benefits in your opinion?

File: tests/phpunit/tests/test-factory.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+		                        ->getMock();
+
+		$message_factory->method( 'create' )->willReturn( $this->createMock( 'WPNotify_Message' ) );
+		$message_factory->method( 'accepts' )->willReturn( true );
+
+		$sender_factory = $this->getMockBuilder( 'WPNotify_SenderFactory' )
+		                       ->setMethods( array( 'create' ) )
+		                       ->getMock();
+
+		$sender_factory->method( 'create' )->willReturn( $this->createMock( 'WPNotify_Sender' ) );
```
Comment (raaaahman) [inline tests/phpunit/tests/test-factory.php:21]: Good. With all these mocks, you won't have to rewrite your test each time we change the code of any of these factories.

File: includes/Factory.php (hunk 1 above)
Comment (danbilauca) [reply includes/Factory.php]: I've fixed it by using array

File: includes/senders/SenderFactory.php (hunk 2 above)
Comment (danbilauca) [reply includes/senders/SenderFactory.php]: yep, fixed it by removing it

**Review by raaaahman — APPROVED** [review]

**Pull request merged**